src/mpid/common/shm/mpidu_shm_alloc.c

```diff
@@ -100,10 +112,15 @@ int MPIDU_shm_seg_alloc(size_t len, void **ptr_p, MPL_memory_class class)
 
     ep->ptr_p = ptr_p;
     ep->len = len;
+    ep->kind = kind;
 
-    ALLOCQ_ENQUEUE(ep);
-
-    segment_len += len;
+    if (kind == MPIR_MEMKIND__DRAM) {
+        GENERIC_Q_ENQUEUE(&allocq_dram, ep, next);
+        dram_segment_len += len;
+    } else {
+        GENERIC_Q_ENQUEUE(&allocq_hybr, ep, next);
+        hybr_segment_len += len;
+    }
```

Comment:
I'm not missing some logic here.  Once the memory is allocated in either `DRAM` or `MCDRAM`, why does the rest of the code need to know about it?

Reply:
After memory has been accounted by `MPIDU_shm_seg_alloc`s it has to be allocated and bound to the corresponding memory type in `MPIDU_shm_seg_commit`. The `allocq_hybr` has to retain the memory type information (`MCDRAM`, `NVRAM`, etc) so that the commit function can perform the right binding. This does not allow for mixed hybrid memory allocations (e.g., `FASTBOXES:MCDRAM` and `CELLS:NVRAM`).